File: structural/dart/composite_structured/components.dart
abstract class FileSystemComponent {
  void display([String indent = '']);
  int getSize();
  
  void add(FileSystemComponent component) { print('Cannot add to a file'); }
  void remove(FileSystemComponent component) { print('Cannot remove from a file'); }
  FileSystemComponent? getChild(int index) { return null; }
}

abstract class Employee {
  void print([String indent = '']);
  void add(Employee employee) { print('Cannot add to a regular employee'); }
  void remove(Employee employee) { print('Cannot remove from a regular employee'); }
}

abstract class Graphic {
  void draw([String indent = '']);
}

File: structural/typescript/composite_structured/components.ts
export abstract class FileSystemComponent {
  abstract display(indent?: string): void;
  abstract getSize(): number;
  
  add(component: FileSystemComponent): void { console.log('Cannot add to a file'); }
  remove(component: FileSystemComponent): void { console.log('Cannot remove from a file'); }
  getChild(index: number): FileSystemComponent | null { return null; }
}

export abstract class Employee {
  abstract print(indent?: string): void;
  add(employee: Employee): void { console.log('Cannot add to a regular employee'); }
  remove(employee: Employee): void { console.log('Cannot remove from a regular employee'); }
}

export abstract class Graphic {
  abstract draw(indent?: string): void;
}
